From 92b37b6902047d74591bf2f6ddbe0ce9f250df4f Mon Sep 17 00:00:00 2001
From: Rajat Soni <quic_rajson@quicinc.com>
Date: Thu, 28 Jul 2022 17:44:56 +0530
Subject: [PATCH] ath12k: Fix memory free during pci power down.

While asserting fw crash for each interface infinite times
one of the interface was not recovered because memory is
not freed during pci power down.

So invoked ath12k_qmi_free_target_mem_chunk() in
ath12k_pci_power_down() to free mem chunk during fw crash.

Signed-off-by: Rajat Soni <quic_rajson@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/pci.c | 3 +++
 drivers/net/wireless/ath/ath12k/qmi.c | 2 +-
 drivers/net/wireless/ath/ath12k/qmi.h | 1 +
 3 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/pci.c b/drivers/net/wireless/ath/ath12k/pci.c
index e92d92a..5845aa7 100644
--- a/drivers/net/wireless/ath/ath12k/pci.c
+++ b/drivers/net/wireless/ath/ath12k/pci.c
@@ -1216,6 +1216,9 @@ void ath12k_pci_power_down(struct ath12k_base *ab)
 {
 	struct ath12k_pci *ab_pci = ath12k_pci_priv(ab);
 
+	if (!(test_bit(ATH12K_FLAG_UNREGISTERING, &ab->dev_flags))) {
+		ath12k_qmi_free_target_mem_chunk(ab);
+	}
 	/* restore aspm in case firmware bootup fails */
 	ath12k_pci_aspm_restore(ab_pci);
 
diff --git a/drivers/net/wireless/ath/ath12k/qmi.c b/drivers/net/wireless/ath/ath12k/qmi.c
index 882f135..9419d5a 100644
--- a/drivers/net/wireless/ath/ath12k/qmi.c
+++ b/drivers/net/wireless/ath/ath12k/qmi.c
@@ -2983,7 +2983,7 @@ out:
 	return ret;
 }
 
-static void ath12k_qmi_free_target_mem_chunk(struct ath12k_base *ab)
+void ath12k_qmi_free_target_mem_chunk(struct ath12k_base *ab)
 {
 	int i;
 
diff --git a/drivers/net/wireless/ath/ath12k/qmi.h b/drivers/net/wireless/ath/ath12k/qmi.h
index 2b3ddab..cfe303f 100644
--- a/drivers/net/wireless/ath/ath12k/qmi.h
+++ b/drivers/net/wireless/ath/ath12k/qmi.h
@@ -747,4 +747,5 @@ int ath12k_send_qdss_trace_mode_req(struct ath12k_base *ab,
 int ath12k_enable_fwlog(struct ath12k_base *ab);
 int ath12k_qmi_m3_dump_upload_done_ind_send(struct ath12k_base *ab,
 					    u32 pdev_id, int status);
+void ath12k_qmi_free_target_mem_chunk(struct ath12k_base *ab);
 #endif
-- 
2.17.1

